File: core/src/main/java/r/base/Namespaces.java
/*
 * R : A Computer Language for Statistical Data Analysis
 * Copyright (C) 1995, 1996  Robert Gentleman and Ross Ihaka
 * Copyright (C) 1997--2008  The R Development Core Team
 * Copyright (C) 2003, 2004  The R Foundation
 * Copyright (C) 2010 bedatadriven
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package r.base;

import org.renjin.primitives.annotations.Current;

import r.lang.*;
import r.lang.exception.EvalException;

public class Namespaces {

  public static SEXP getRegisteredNamespace(@Current Context context, String name) {
    return getRegisteredNamespace(context, Symbol.get(name));
  }

  public static SEXP getRegisteredNamespace(@Current Context context, Symbol name) {
    return context.findNamespace(name);
  }

  public static void registerNamespace(@Current Context context, String name, Environment env) {
    Frame registry = context.getGlobals().namespaceRegistry;
    Symbol symbol = Symbol.get(name);
    if(registry.getVariable(symbol) != Symbol.UNBOUND_VALUE) {
      throw new EvalException("name space already registered");
    }
    registry.setVariable(symbol, env);
  }

  public static boolean isNamespaceEnv(@Current Context context, SEXP envExp) {
    if(envExp == context.getGlobals().baseNamespaceEnv) {
      return true;
    } else if(envExp instanceof Environment) {
      Environment env = (Environment)envExp;
      SEXP info = env.getVariable(Symbol.get(".__NAMESPACE__."));
      if(info instanceof Environment) {
        SEXP spec = ((Environment)info).getVariable(Symbol.get("spec"));
        if(spec instanceof StringVector && spec.length() > 0) {
          return true;
        } else {
          return false;
        }
      }
    }
    return false;
  }

  public static void importIntoEnv(Environment impenv, Vector impnames, Environment expenv, Vector expnames) {

    /* This function copies values of variables from one environment
       to another environment, possibly with different names.
       Promises are not forced and active bindings are preserved. */

    if(impnames.length() != expnames.length()) {
      throw new EvalException("length of import and export names must match");
    }

    for(int i=0;i!=impnames.length();++i) {
      if(impnames.isElementNA(i) || expnames.isElementNA(i)) {
        throw new EvalException("Import/export name cannot be NA");
      }
      Symbol impsym = Symbol.get(impnames.getElementAsString(i));
      Symbol expsym = Symbol.get(expnames.getElementAsString(i));

      SEXP value = expenv.findVariable(impsym);
      impenv.setVariable(expsym, value);
    }
  }

}

File: tests/src/test/R/namespaces.R
#
# Renjin : JVM-based interpreter for the R language for the statistical analysis
# Copyright © 2010-2018 BeDataDriven Groep B.V. and contributors
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, a copy is available at
# https://www.gnu.org/licenses/gpl-2.0.txt
#

library(hamcrest)
library(stats)
library('org.renjin.test:thirdparty')

test.getNamespaceName <- function() {
    
    assertThat(getNamespaceName(environment(rnorm)), identicalTo(c(name = "stats")))
    assertThat(getNamespaceName(environment(qr)), identicalTo("base"))
}

test.getNamespaceNameOfThirdParty <- function() {
    assertThat(getNamespaceName(environment(compute)), identicalTo(c(name = "org.renjin.test:thirdparty")))
}

test.getNamespaceExports <- function() {
    
    assertTrue("model.matrix.lm" %in% getNamespaceExports(environment(rnorm)))
    assertTrue("model.matrix.lm" %in% getNamespaceExports("stats"))
    assertTrue("model.matrix.lm" %in% getNamespaceInfo(environment(rnorm), which = "exports"))
    assertTrue("for" %in% getNamespaceExports(environment(qr)))
    assertTrue("for" %in% getNamespaceExports("base"))
}

test.isNamespaceLoaded <- function() {
    assertTrue(isNamespaceLoaded("stats"))
}

# Not yet implemented
ignore.test.getNamespaceImports <- function() {

    imports <- getNamespaceImports("stats")
    print(imports)
    assertThat(typeof(imports), identicalTo("list"))
    assertThat(imports$base, identicalTo(TRUE))
    assertThat(imports$utils, identicalTo(str = str))
}

test.requireNamespace <- function() {
    assertTrue(requireNamespace("stats"))
}

test.requireNamespaceNonExistant <- function() {
    assertFalse(requireNamespace("fooooobarrr"))
}

test.library.multiple <- function() {
    searchPathLen <- length(search())
    
    library(stats)
    library(stats)
    
    assertThat(length(search()), identicalTo(searchPathLen))
}